blockchain/chain_sync/src/network_context.rs

```diff
@@ -172,35 +175,69 @@ where
                 // a request succeeds.
                 let peers = self.peer_manager.top_peers_shuffled().await;
-                let mut res = None;
+                let mut tasks = JoinSet::new();
                 for p in peers.into_iter() {
-                    match self.chain_exchange_request(p, request.clone()).await {
-                        Ok(bs_res) => match bs_res.into_result() {
-                            Ok(r) => {
-                                res = Some(r);
-                                break;
-                            }
-                            Err(e) => {
-                                debug!("Failed chain_exchange response: {}", e);
-                                continue;
+                    let ctl_tx = ctl_tx.clone();
+                    let ctl_rx = ctl_rx.clone();
+                    let res_tx = res_tx.clone();
+                    let peer_manager = self.peer_manager.clone();
+                    let network_send = self.network_send.clone();
+                    let request = request.clone();
+                    tasks.spawn(async move {
+                        if ctl_tx.send_async(()).await.is_ok() {
+                            match Self::chain_exchange_request(
+                                peer_manager,
+                                network_send,
+                                p,
+                                request,
+                            )
+                            .await
+                            {
+                                Ok(bs_res) => match bs_res.into_result() {
+                                    Ok(r) => {
+                                        _ = res_tx.send_async(r).await;
+                                    }
+                                    Err(e) => {
+                                        _ = ctl_rx.recv_async().await;
+                                        debug!("Failed chain_exchange response: {}", e);
+                                    }
+                                },
+                                Err(e) => {
+                                    _ = ctl_rx.recv_async().await;
+                                    debug!("Failed chain_exchange request to peer {:?}: {}", p, e);
+                                }
                             }
-                        },
-                        Err(e) => {
-                            debug!("Failed chain_exchange request to peer {:?}: {}", p, e);
-                            continue;
                         }
-                    }
+                    });
+                }
+
+                async fn wait_all<T: 'static>(tasks: &mut JoinSet<T>) {
+                    while tasks.join_next().await.is_some() {}
                 }
 
-                res.ok_or_else(|| "ChainExchange request failed for all top peers".to_string())?
+                tokio::select! {
+                    r = res_rx.recv_async() => {
```

Comment:
let's use full names for variables; I am not sure what `r` means (I guess `result`?). Rust is complicated enough; let's not add any additional mental overhead. The same goes for, e.g. `ctl_rx` above.

Reply:
Fixed